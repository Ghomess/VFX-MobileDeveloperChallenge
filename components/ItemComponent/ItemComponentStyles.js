import {StyleSheet} from 'react-native';
import {colors} from '../../styles';

export const itemComponentStyles = StyleSheet.create({
  container: {
    borderRadius: 5,
    padding: 15,
    paddingLeft: 15,
  },
  text: {
    color: colors.white,
    fontSize: 15,
  },
});
