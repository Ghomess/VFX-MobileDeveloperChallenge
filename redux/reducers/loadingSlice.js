import {createSlice} from '@reduxjs/toolkit';

const initialState = {
  loadingChart: false,
  loadingSearchResults: false,
};

const loadingSlice = createSlice({
  name: 'loading',
  initialState,
  reducers: {
    changeLoadingChart: (state, action) => {
      console.log('CHANGE loadingChart: ', action.payload);
      state.loadingChart = action.payload;
    },
    changeLoadingSearchResults: (state, action) => {
      console.log('CHANGE loadingSearchResults: ', action.payload);
      state.loadingSearchResults = action.payload;
    },
  },
});

export const {changeLoadingChart, changeLoadingSearchResults} =
  loadingSlice.actions;

export default loadingSlice.reducer;
