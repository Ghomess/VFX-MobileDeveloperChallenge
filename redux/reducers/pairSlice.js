import {createSlice} from '@reduxjs/toolkit';

const initialState = {
  pair: null,
  pairData: [],
  pairDateSelected: '',
  pairDateType: 'Daily',
};

const pairSlice = createSlice({
  name: 'pair',
  initialState,
  reducers: {
    addPair: (state, action) => {
      console.log('ADD PAIR: ', action.payload);
      state.pair = action.payload;
    },
    addPairData: (state, action) => {
      console.log('ADD PAIR DATA: ', action.payload);
      state.pairData = action.payload;
    },
    addPairDateSelected: (state, action) => {
      console.log('ADD PAIR DATE: ', action.payload);
      state.pairDateSelected = action.payload;
    },
    addPairDateType: (state, action) => {
      console.log('ADD PAIR DATE: ', action.payload);
      state.pairDateType = action.payload;
    },
  },
});

export const {addPair, addPairData, addPairDateSelected, addPairDateType} =
  pairSlice.actions;

export default pairSlice.reducer;
