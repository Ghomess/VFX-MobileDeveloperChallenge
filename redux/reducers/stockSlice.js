import {createSlice} from '@reduxjs/toolkit';

const initialState = {
  ticker: null,
  stockData: {},
  tickerSearchResults: [],
  stockMonthSelected: null,
  stockDateType: 'Monthly',
  stockPrice: null,
};

const stockSlice = createSlice({
  name: 'ticker',
  initialState,
  reducers: {
    addTicker: (state, action) => {
      console.log('ADD ticker: ', action.payload);
      state.ticker = action.payload;
    },
    addStockData: (state, action) => {
      console.log('ADD stock DATA: ', action.payload);
      state.stockData = action.payload;
    },
    addStockMonthSelected: (state, action) => {
      console.log('ADD stock MONTH: ', action.payload);
      state.stockMonthSelected = action.payload;
    },
    addTickerSearchResults: (state, action) => {
      console.log('ADD ticker SEARCH RESULTS: ', action.payload);
      state.tickerSearchResults = action.payload;
    },
    addStockPrice: (state, action) => {
      console.log('ADD stock PRICE: ', action.payload);
      state.stockPrice = action.payload;
    },
  },
});

export const {
  addTicker,
  addStockData,
  addStockMonthSelected,
  addTickerSearchResults,
  addStockPrice,
} = stockSlice.actions;

export default stockSlice.reducer;
