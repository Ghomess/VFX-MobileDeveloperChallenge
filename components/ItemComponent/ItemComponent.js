import React from 'react';
import {Text, TouchableOpacity} from 'react-native';

import {useDispatch} from 'react-redux';
import {addTicker} from '../../redux/reducers/stockSlice';
import {addPair} from '../../redux/reducers/pairSlice';

import {changeLoadingChart} from '../../redux/reducers/loadingSlice';
import {itemComponentStyles} from './ItemComponentStyles';

export const ItemComponent = ({name, setEnableList, screen}) => {
  const dispatch = useDispatch();
  const handlePress = () => {
    if (screen === 'Stocks') {
      dispatch(addTicker(name));
    } else if (screen === 'Currency') {
      dispatch(addPair(name));
    }
    dispatch(changeLoadingChart(true));
    setEnableList(false);
  };
  return (
    <TouchableOpacity
      style={itemComponentStyles.container}
      onPress={handlePress}
      testID="ItemComponent.ToucableOpacity">
      <Text style={itemComponentStyles.text}>{name}</Text>
    </TouchableOpacity>
  );
};
