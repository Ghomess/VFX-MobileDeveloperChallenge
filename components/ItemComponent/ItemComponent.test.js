import React from 'react';

import {NavigationContainer} from '@react-navigation/native';
import {Provider} from 'react-redux';

import {fireEvent, render, waitFor} from '@testing-library/react-native';

import {ItemComponent} from './ItemComponent';
import stockSlice, {addTicker} from '../../redux/reducers/stockSlice';
import {configureStore} from '@reduxjs/toolkit';
import pairSlice, {addPair} from '../../redux/reducers/pairSlice';
import loadingSlice, {
  changeLoadingChart,
} from '../../redux/reducers/loadingSlice';
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    getState: () => ({
      routes: [{name: 'Stocks'}, {name: 'Currency'}],
      index: 0,
    }),
  }),
}));

function createTestStore() {
  const store = configureStore({
    reducer: {
      ticker: stockSlice,
      pair: pairSlice,
      loading: loadingSlice,
    },
  });
  return store;
}

let store;
describe('ItemComponent', () => {
  beforeEach(() => {
    store = createTestStore();
  });
  test('renders correctly in stock screen', async () => {
    await waitFor(() => {
      const tree = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'META'}
              screen={'Stocks'}
              setEnableList={jest.fn()}
            />
          </NavigationContainer>
        </Provider>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  test('renders correctly in the currency screen', async () => {
    await waitFor(() => {
      const tree = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'EUR/USD'}
              screen={'Currency'}
              setEnableList={jest.fn()}
            />
          </NavigationContainer>
        </Provider>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  test('check if it calls setEnableList and changes it to false on press in stock screen', async () => {
    await waitFor(() => {
      const setEnableListMock = jest.fn();

      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'META'}
              screen={'Stocks'}
              setEnableList={setEnableListMock}
            />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemComponent.ToucableOpacity');
      fireEvent.press(touchable);
      expect(setEnableListMock).toHaveBeenCalledTimes(1);
      expect(setEnableListMock).toHaveBeenCalledWith(false);
    });
  });

  test('check if it calls setEnableList and changes it to false on press in currency screen', async () => {
    await waitFor(() => {
      const setEnableListMock = jest.fn();

      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'EUR/USD'}
              screen={'Currency'}
              setEnableList={setEnableListMock}
            />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemComponent.ToucableOpacity');
      fireEvent.press(touchable);
      expect(setEnableListMock).toHaveBeenCalledTimes(1);
      expect(setEnableListMock).toHaveBeenCalledWith(false);
    });
  });

  //Dispatches

  test('check if it dispatches addTicker action in stocks screen', async () => {
    await waitFor(() => {
      const dispatchMock = jest.spyOn(store, 'dispatch');
      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'META'}
              screen={'Stocks'}
              setEnableList={() => {}}
            />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemComponent.ToucableOpacity');
      fireEvent.press(touchable);
      expect(dispatchMock).toHaveBeenCalledWith(addTicker('META'));
    });
  });

  test('check if it dispatches addPair action in currency screen', async () => {
    await waitFor(() => {
      const dispatchMock = jest.spyOn(store, 'dispatch');
      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'EUR/USD'}
              screen={'Currency'}
              setEnableList={() => {}}
            />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemComponent.ToucableOpacity');
      fireEvent.press(touchable);
      expect(dispatchMock).toHaveBeenCalledWith(addPair('EUR/USD'));
    });
  });

  test('check if it dispatches changeLoadingChart action in currency screen', async () => {
    await waitFor(() => {
      const dispatchMock = jest.spyOn(store, 'dispatch');
      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'EUR/USD'}
              screen={'Currency'}
              setEnableList={() => {}}
            />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemComponent.ToucableOpacity');
      fireEvent.press(touchable);
      expect(dispatchMock).toHaveBeenCalledWith(changeLoadingChart(true));
    });
  });
  test('check if it dispatches changeLoadingChart action in stock screen', async () => {
    await waitFor(() => {
      const dispatchMock = jest.spyOn(store, 'dispatch');
      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemComponent
              name={'META'}
              screen={'Stocks'}
              setEnableList={() => {}}
            />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemComponent.ToucableOpacity');
      fireEvent.press(touchable);
      expect(dispatchMock).toHaveBeenCalledWith(changeLoadingChart(true));
    });
  });
});
