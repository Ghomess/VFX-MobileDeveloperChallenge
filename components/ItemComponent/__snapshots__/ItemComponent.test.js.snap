// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemComponent renders correctly in stock screen 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "borderRadius": 5,
      "opacity": 1,
      "padding": 15,
      "paddingLeft": 15,
    }
  }
  testID="ItemComponent.ToucableOpacity"
>
  <Text
    style={
      {
        "color": "#ffffff",
        "fontSize": 15,
      }
    }
  >
    META
  </Text>
</View>
`;

exports[`ItemComponent renders correctly in the currency screen 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "borderRadius": 5,
      "opacity": 1,
      "padding": 15,
      "paddingLeft": 15,
    }
  }
  testID="ItemComponent.ToucableOpacity"
>
  <Text
    style={
      {
        "color": "#ffffff",
        "fontSize": 15,
      }
    }
  >
    EUR/USD
  </Text>
</View>
`;
